docs/sender.rst

```diff
@@ -0,0 +1,409 @@
+            at=TimestampNanos.now())
+        sender.flush()
+
+Note that the last `sender.flush()` is entirely optional as flushing
+also happens at the end of the ``with`` block.
+
+.. _sender-auto-flushing:
+
+Auto-flushing
+-------------
+
+To avoid accumulating very large buffers, the sender will - by default -
+flush the buffer automatically.
+
+Auto-flushing is triggered when appending a row to the internal sender
+buffer and the buffer either:
+
+* Reaches 75'000 rows (for HTTP) or 600 rows (for TCP).
+* Hasn't been flushed for 1 second.
+
+Here is an example that auto-flushes every 10 rows and disables
+the interval-based auto-flushing logic.
+
+``http::addr=localhost:9000;auto_flush_rows=10;auto_flush_interval=off;``
+
+Here is auto-flushing disabled:
+
+``http::addr=localhost:9000;auto_flush=off;``
+
+See the :ref:`configuring-sender-auto-flushing` section for more details.
+
+Error Reporting
+===============
+
+**TL;DR: Use HTTP for better error reporting**
+
+The sender will do its best to check for errors before sending data to the
+server.
+
+When using the HTTP protocol, the server will send back an error message if
+the data is invalid or if there is a problem with the server. This will be
+raised as an :class:`questdb.ingress.IngressError` exception.
+
+The HTTP layer will also attempt retries, configurable via the 
+:ref:`retry_timeout <configuring-sender-request>` parameter.`
+
+When using the TCP protocol errors are *not* sent back from the server and
+must be searched for in the logs. See the :ref:`troubleshooting-flushing`
+section for more details.
+
+HTTP transactions
+=================
+
+When using the HTTP protocol, the sender can be configured to send a batch of
+rows as a single transaction.
+
+**Transactions are limited to a single table.**
+
+If auto-flusing is enabled, any pending data will be flushed before the
+transaction is started.
+
+Auto-flushing is disabled during the scope of the transaction.
+
+The transaction is automatically committed when the ``with`` block is exited.
+
+.. code-block:: python
+
+    conf = 'http::addr=localhost:9000;'
+    with Sender.from_conf(conf) as sender:
+        with sender.transaction('weather_sensor') as txn:
+            txn.row(
+                symbols={'id': 'toronto1'},
+                columns={'temperature': 23.5, 'humidity': 0.49},
+                at=TimestampNanos.now())
+            txn.row(
+                symbols={'id': 'dubai2'},
+                columns={'temperature': 41.2, 'humidity': 0.34},
+                at=TimestampNanos.now())
+
+You can complete a transaction explicity by calling the
+:func:`questdb.ingress.SenderTransaction.commit` or the
+:func:`questdb.ingress.SenderTransaction.rollback` methods.
+
+Raising an exception from within the transaction ``with`` block will also cause
+the transaction to be rolled back.
+
+
+.. _configuring-sender:
+
+Full configuration options
+==========================
+
```

Comment:
Re-introduce the conf string format basics and provide a few more examples.

Reply:
The Rust, C, C++ and Python APIs are consistent between the builder APIs and the configuration format.

This was a pain to implement, but it's not something that will be done with other clients where their APIs will diverge from the config.

At least the config should be consistent across.